### ISSUE TITLE
Pure Rust implementation

### ISSUE THREAD

**Issue by adeschamps** [adeschamps/lcm#28]
In #24 we talked about writing a pure Rust implementation of LCM. Motivations include:

- Proper handling of panics
- Not having to pay the cost of a mutex in a single threaded environment
- Possibly making use of tokio

Handling panics properly could probably be done by imroving the current implementation, while the other two cannot.

Another reason why a pure Rust implementation would be nice is that it removes the dependency on a system library.

That said, reimplementing LCM isn't going to be trivial. We should probably list out all the things that would need to be done before committing to it.

**Comment by neachdainn** [comment]
I think the first thing we need to do is decide how much this implementation needs to behave like the C/C++ version:

* The C version is thread-safe, so I assume some LCM users do use multithreaded LCM. I propose we *don't* worry about making it thread safe and instead do one of:
    * Make an LCM client incredibly cheap to create. Perhaps by having a single LCM backend (e.g., the Tokio part) which connects to multiple `Lcm` objects.
    * Write it in such a way that using a `Mutex<Lcm>` is very easy. I am not sure yet how this will restrict the callback functions, if at all (e.g., will they have to be `'static`).
    * Totally ignore the issue until someone actually claims to need it.
* Do we need to be able to adjust the queue size dynamically? If we only allow the queue size to be specified upon subscribing to the channel, we could potentially do some optimizations. For example, we could use a lock-free array based SPSC verses a linked list or a lock based queue.
    * I have a potential implementation for such a queue [here](https://gist.github.com/neachdainn/ac8958f598986657712466cba3b63315). It's probably not very good but it's what I came up with while playing around with this problem.
* We could also go radically different and not use a callback system. Instead, subscribing could expose the receiving end of a queue. Unsubscribing from a channel would be as simple as dropping the receiver.
    * I really doubt we want to do this but it could potentially tie into other libraries in interesting ways and it would solve the multithreading issue.
* Do we want to try and make LCM compatible with other parts of the Rust ecosystem?
    * Make this based on Serde and write a `serde_lcm` crate?
    * Make it easy to tie into [futures](https://docs.rs/futures/*/futures/stream/trait.Stream.html)?
* Etc

**Comment by neachdainn** [comment]
Note that panics across FFI boundaries are not longer undefined behavior as of Rust 1.24 (see [here](https://github.com/rust-lang/rust/pull/46833) and [here](https://blog.rust-lang.org/2018/02/15/Rust-1.24.html)).

Additionally, Tokio is going through a [reform](https://tokio.rs/blog/2018-02-tokio-reform-shipped/).  I'm not sure what that means about if/when we should start this refactoring.

**Comment by adeschamps** [comment]
That's exciting news!

I like the idea of putting messages in a queue instead of a callback mechanism, but I wonder if that's actually the right design choice. If an API offers callbacks but you want a queue, it's easy to write simple callbacks that just push messages onto a queue. On the other hand, if an API offers queues but you want callbacks, there's more work involved.

On the threading issue, have you ever encountered a situation where `lcm_handle` was called from more than one thread? Assuming a callback API, if the callbacks take ownership of messages, then it's easy to just push them onto a threadsafe queue if need be.

I may be wrong, but I don't think LCM serialization can be implemented with serde, for similar reasons that prost couldn't. The lengths of arrays are specified by the values of other fields, in a way that serde doesn't have the ability to express.

**Comment by neachdainn** [comment]
I agree on the queue vs callback system. It was an interesting thought but probably not the right direction to go.

In my own personal use, I have never encountered that situation but I believe the issue would be less about sending the messages across threads and more about sending the callback across threads. I may be totally wrong about this, but I think having the callbacks run on multiple threads change the callback from `FnMut(T) + 'a` to `Fn(T) + Sync + Send + 'static`, even in a single-threaded (from the user perspective) context.

I.e.,
```rust
let mut a = 0;
let lcm = Lcm::new();
lcm.subscribe("channel", |m| a = m);
``` 

vs

```rust
let a = Arc::new(Mutex::new(0));
let a_clone = a.clone();

let lcm = Lcm::new();
lcm.subscribe("channel", move |b| *a_clone.lock() = b);
```
